doc/users-guide/datapar/promotion.rst

```diff
@@ -0,0 +1,324 @@
+:title: Users Guide: Promotion
+
+.. _ug-promotion:
+
+Promotion: implicit data parallelism
+====================================
+
+In Chapel, *function promotion* (or simply *promotion*) refers to an
+implicit form of data parallelism that is triggered by passing a
+collection of values to an argument that expects a single value.  More
+precisely, when an array of type *t* (or a range or domain whose index
+type is *t*) is passed to a function argument of type *t*, that
```

Comment:
I added issue #5114 for this and filed some tests against.   In doing so, I learned that the fragility I referred to yesterday was a general fragility in resolving parallel loops and nothing specific to promotion.  It also turned out that all forall expressions promote, not just those over ranges, domains, and arrays (because I had a bug in my user-defined iterators, I was running into the general forall loop fragility and not anything specific to promotion).

I've also reworded (and reformatted) that sentence to try and list these three cases explicitly.

I agree that there should be a section on forall expressions in the users guide (along with all other language features).  I'm trying to prioritize based on importance / uniqueness to Chapel.  I've added an item to my TODO list to add forall expressions between the existing two data parallel sections.